File: scala-3/ninety-nine/src/main/scala/com/rbleggi/ninetynine/P03.scala
package com.rbleggi.ninetynine

import scala.annotation.tailrec

object P03 {
  @tailrec
  def nth[A](k: Int, items: List[A]): A =
    (k, items) match
      case (k, Nil) => throw new NoSuchElementException("nth of empty list")
      case (k, head :: tail) if k < 0 => throw new NoSuchElementException("negative index")
      case (0, head :: tail) => head
      case (k, head :: tail) => nth(k - 1, tail)
}

@main def mainP03(): Unit =
  println("Find the Kth element of a list")
  val numbers = List(1, 1, 2, 3, 5, 8)
  println(s"The list is: $numbers")
  val k = 4
  println(s"The index to find is: $k")
  val kthElement = P03.nth(k, numbers)
  println(s"The element at index $k is: $kthElement")

File: kotlin/ninety-nine/src/main/kotlin/com/rbleggi/ninetynine/P03.kt
package com.rbleggi.ninetynine

object P03 {
    tailrec fun <A> nth(k: Int, items: List<A>): A {
        println("Recursion step: k = $k, $items")
        return when {
            items.isEmpty() -> throw NoSuchElementException("nth of empty list")
            k < 0 -> throw NoSuchElementException("negative index")
            k == 0 -> items[0]
            else -> nth(k - 1, items.drop(1))
        }
    }
}

fun mainP03() {
    println("Find the Kth element of a list")
    val numbers = listOf(1, 1, 2, 3, 5, 8)
    println("The list is: $numbers")
    val k = 4
    println("The index to find is: $k")
    val kthElement = P03.nth(k, numbers)
    println("The element at index $k is: $kthElement")
}
